### DISCUSSION TITLE
Is it possible to get Where-Object to ignore virtual terminal sequences?

### DISCUSSION THREAD

**Discussion by nmbell** [PowerShell/PowerShell#15733] [Enhancements]
I find virtual terminal sequences very useful for coloring string output, but it also prevents filtering on that property, for example (using the VTS for 'negative' text here to illustrate the example, but of course in the real world the output may use an arbitrary range of VTS's): 

Works with plain text:
```
'Red','Yellow','Green' | ForEach-Object { [PSCustomObject]@{ Color = "$_" } } | Where-Object Color -ne 'Red'
```
```
Color
-----
Yellow
Green
```

Doesn't work with a VTS:
```
'Red','Yellow','Green' | ForEach-Object { [PSCustomObject]@{ Color = "`e[7m$_`e[0m" } } | Where-Object Color -ne 'Red'
```
```
Color
-----
Red
Yellow
Green
```

Unless I happen to know the VTS that was used:
```
'Red','Yellow','Green' | ForEach-Object { [PSCustomObject]@{ Color = "`e[7m$_`e[0m" } } | Where-Object Color -ne "`e[7mRed`e[0m"
```
```
Color
-----
Yellow
Green
```

If `Where-Object` was able to ignore the VTS's, then I could add colors to my output without breaking downstream operations. I was just wondering if it's possible?

**Comment by iSazonov** [comment]
We still do not have a consistent design for coloring and supporting VTS. https://github.com/PowerShell/PowerShell/issues/10780
Although $PSStyle was added there (see #13758 #13071) In my opinion, this is more of a long workaround than a viable solution.

**Comment by AltitudeApps** [comment]
It is with deep shame that I offer a suggestion involving regular expressions:

```pwsh
'Red','Yellow','Green'  `
  |                     `
    ForEach-Object      `
      {                 `
        [PSCustomObject] @{
           Color =
             "`e[7m$_`e[0m"
        };              `
      }                 `
  |                     `
    Where-Object        `
      -Property Color   `
      -NotMatch '.*Red.*'
```

results in:

```txt
Color
-----
Yellow
Green
```

**Reply by nmbell** [reply]
@AltitudeApps Your regex works here, but may be too broad for general use. I could use something like:
```
Where-Object { ($_.Color -replace "`e\[(\d+;?){0,16}m",'') -ne 'Red' }
```
but my point is that if it was included in Where-Object natively, we wouldn't need any such cumbersome constructions.

**Reply by AltitudeApps** [reply]
Understood.

However, my view is that `Where-Object` has been designed to be as general as possible, and I think it would be counter to its design goals to include parameters which target specific use cases.  Where would you stop?

The ability to target specific use cases is satisfied by the option to pass a filter script, expressed as a scriptblock literal, as an argument to `Where-Object`, just as you did in your example. (So long as that scriptblock, when invoked, evaluates to a boolean value.) 

But it does not have to be a scriptblock literal, it can be any expression which evaluates to a scriptblock with a single clause, including a variable which is pointing to a scriptblock which was created sometime earlier in program execution.

So, sometimes what I like to do is to create these as scriptblocks in advance, so that they are less of an "ad-hoc" construct, and also so that I can reuse them without "cat on a keyboard" code in the middle of something larger.

```pwsh
$sbFilterVTS = `
  {
    process {
      ($_.Color -replace "`e\[(\d+;?){0,16}m",'') -ne $colorStr
    }
  }

$sbFilterVTSRed = & { $colorStr = 'Red'; $sbFilterVTS.GetNewClosure() }

'Red','Yellow','Green' `
  | `
    ForEach-Object {
      [PSCustomObject]@{
        Color = "`e[7m$_`e[0m"
      }
    } `
  | `
    Where-Object $sbFilterVTSRed
```
or maybe
```pwsh
$sbFilterVTSB = `
  {
    process {
      -not $_.Color.Contains( $colorStr )
    }
  }

$sbFilterVTSRedB = & { $colorStr = 'Red'; $sbFilterVTSB.GetNewClosure() }

'Red','Yellow','Green' `
  | `
    ForEach-Object {
      [PSCustomObject]@{
        Color = "`e[7m$_`e[0m"
      }
    } `
  | `
    Where-Object $sbFilterVTSRedB
```
---------------------
Well, that was a long trip, only to get to a simpler point:

In your original post, you were using the `-ne` parameter, which corresponds to the `-ne` operator.  But the `-notcontains` parameter also exists:

```pwsh
'Red','YellowE','GreenE','BlueE','CyanE' `
  | `
    ForEach-Object {
      [PSCustomObject]@{
        Color = "`e[7m$_`e[0m"
      }
    } `
  | `
    Where-Object -Property Color -NotContains 'Red'
    #  not too cumbersome, I hope
```
This will do what you asked for in the original question.  (I think.)  But it is likely that your use case involves things where the filtering needs to be more sophisticated than what is possible by using any of the "specify use of this operator" style of parameters that `Where-Object` offers.  I almost always end up using a filter script anyway, in spite of the wide array of options available with `Where-Object`.

**Reply by nmbell** [reply]
Your first two examples work, but are even more cumbersome than the regex I gave earlier. Your last example still returns `Red` in the output. I realize there are many different ways that this *could* be achieved, but I don't think it should be necessary. Imagine I write a function that produces an arbitrary number of output objects. You use the function, but need to limit it to a subset of the results. You do so using a simple -ne filter. I now decide to add some arbitrary coloring to the output. Your script now breaks. You now have to add some custom filtering to account for the fact that I added colors to my output, even though the *content* of the output has not changed. My use case is simply that I'd like to be able to filter based on the characters that appear on the screen, without having to account for invisible VTS codes. Being able to color text is great, but, imho, it should be transparent to any downstream code.

**Reply by nmbell** [reply]
I also just discovered that the VTS's are persisted when sending output to a file, which basically rules them out for anything you might ever want to write to a file, which is basically everything, lol.
